File: <100/014/014.cpp
#include<iostream>
using namespace std;
typedef int64_t ll;
#define MAX 1000000

ll answer, cnt[MAX+10], max_cnt;
const ll upper_limit = MAX;

ll Collatz(ll n){
  if(n < upper_limit && 0 < cnt[n]){
    return cnt[n];
  }

  ll m = n;
  if(m&1) m = 3*m + 1;
  else    m = m/2;
  ll C = Collatz(m)+1;

  if(n < upper_limit) cnt[n] = C;
  return C;
}

int main(){
  cnt[1] = 1;

  for(ll i = 1; i < upper_limit; ++i){
    ll C = Collatz(i);
    if(max_cnt < C){
      max_cnt = C;
      answer = i;
    }
  }

  cout << answer << endl;
  return 0;
}

File: <100/014/014.py
UPPER_LIMIT: int = 10**6
cnt = [0 for _ in range(10**6+10)]


def Collatz(n: int) -> int:
    c = 0
    m = n

    while True:
        if m < UPPER_LIMIT and cnt[m] > 0:
            c += cnt[m]
            break
        c += 1

        if m%2 == 0:
            m = m // 2
        else:
            m = 3*m + 1
    cnt[n] = c

    return c

def main() -> None:
    chain_l: int = 0
    answer: int  = 0
    cnt[1] = 1

    for j in range(1, UPPER_LIMIT):
        c = Collatz(j)
        if chain_l < c:
            chain_l = c
            answer = j

    print(answer)

    return


if __name__ == '__main__':
    main()

